Implement "go to definition" for `render` calls in ERB templates

janko opened the pull request:
It's common to want to traverse through several partials while updating HTML that a controller action renders. Rails.vim has a neat `gf` shortcut for this. This brings the same functionality to Ruby LSP Rails, by implementing "go to definition" support for `render` calls inside ERB templates.

https://github.com/user-attachments/assets/51080039-4655-4de5-84d1-4307da8a1802

It supports partial name passed as positional argument, or via `:partial`, `:layout`, and `:spacer_template` keyword arguments. It even handles `:variants`, `:formats`, and `:handlers` options, as well as `:template` for rendering non-partial templates.

The initial implementation mimicked the template lookup logic inside the language server. However, that turned out to be more complex and didn't handle custom `view_paths`. So, I ended up calling `ActionView::LookupContext` to perform the actual template lookup on the server. I also needed to determine the controller from the template directory in a way that handles custom `view_paths`, as well as handle controllerless template directories.

To avoid the overhead of booting the Rails process too many times in tests, I updated the test helpers to allow sending multiple `textDocument/definition` requests to the same server.

janko committed:
Implement "go to definition" for render calls in ERB templates

It's common to want to traverse through several partials while updating
HTML that a controller action renders. Rails.vim has a neat `gf`
shortcut for this, though it probably doesn't have the precision that
Prism would provide.

This brings the same functionality to Ruby LSP Rails, by implementing
"go to definition" support for render calls inside ERB templates. It
supports partial name passed as positional argument, or via `:partial`,
`:layout`, and `:spacer_template` keyword arguments. It even handles
`:variants`, `:formats`, and `:handlers` options, as well as `:template`
for rendering non-partial templates. Relative lookup will also check
in view directories of controller ancestors.

For the latter, I considered doing a call to the Rails process that will
return `ActionController::Base._prefixes`. However, I couldn't think of
a good enough interface, and that method ableit public is undocumented,
so it seems like we shouldn't rely on it. Given that this ancestry
lookup is non-configurable anyway, I chose to implement it in Ruby LSP
land based on indexed controller files.

To avoid the overhead of booting the Rails process too many times in
tests, I updated the test helpers to allow sending multiple
`textDocument/definition` requests to the same server.

rafaelfranca reviewed (commented):
Made a few comments. Typechecker is failing and new methods are missing typing information.

janko commented on lib/ruby_lsp/ruby_lsp_rails/runner_client.rb:
```diff
@@ -261,6 +261,10 @@ def connected?
       end
 
+      def views_dir
+        File.join(@rails_root, "app/views")
```
For now, I would just retrieve `ActionController::Base.view_paths`, and hope it would cover the most common cases. Will that work in the Shopify monolith?

janko committed:
Make template lookup more robust by calling server

Also add missing type annotations.

janko commented on lib/ruby_lsp/ruby_lsp_rails/definition.rb:
```diff
@@ -142,6 +147,47 @@ def handle_association(node)
         @response_builder << Support::LocationBuilder.line_location_from_s(result.fetch(:location))
       end
 
+      def handle_possible_render(node)
```
I moved the template lookup into the server, by actually performing it on `ActionView::LookupContext`. I realized that's simpler and more reliable than reimplementing it in the language server. Now the method is hopefully simple enough.

janko committed:
Handle custom view paths of controller subclasses

While here, we undo changes in code lens to handle custom view paths. If
this approach is accepted, we can always update the code lens later.

janko commented:
@rafaelfranca I believe I addressed all your feedback and fixed typechecking.

rafaelfranca reviewed (approved):
Amazing work!

@Shopify/ruby-dx can someone do a second review?

janko committed:
Handle render calls from controllerless template directories

In our app, we have an `app/views/shared/components` directory that
contains shared "component" partials. It doesn't have any matching
controller, and components can render other components, you just need
to specify the full path to the render call (we have a helper method for
that). We should make go to definition work for these cases as well.

leonvogt commented:
This would be awesome and such a quality of life improvement!

janko committed:
Merge branch 'main' into go-to-render

janko committed:
Support only keyword argument passes as the first argument

We want to support `render partial: "foo"` but not `render "foo",
partial: "bar"`, because in the latter case, `:partial` will be a
template local.
